recipe_scrapers/__init__.py

```diff
@@ -605,22 +607,40 @@ def scraper_exists_for(url_path: str) -> bool:
     return host_name in get_supported_urls()
 
 
+def check_scraper_allowed(url_path, host_name, **options: Any) -> bool:
+    """Checks the website's robots.txt file to see if scraping is allowed.
+    If scraping is allowed, returns true. Note that this does not check to see
+    when the last time robots.txt was checked, so this could lead to constant pings
+    on robots.txt"""
```

Comment:
I think that this is OK - because for local/hobby use, the amount of request traffic should be relatively low.  Even so, thank you for making it clear in the comment.

Reply:
Hrm.  Well, is it really OK?  If we can avoid sending unnecessary requests, we should do - even at small scale.  I'll think about this some more.

Reply:
Unfortunately I think I'm changing my opinion here, and that we should be stricter and attempt to cache the results of these `robots.txt` retrievals.  That means more work up-front, but given the relatively small timedelta it introduces between the longer-timescales of {past: no builtin robots txt retrieval per session} and potential {future: one-or-lots-of automatic builtin txt retrievals per scrape}, I think it's worth the time here to navigate towards the moderated-requests end of the spectrum.

Reply:
Hmm...  any ideas on the best way to implement this? I definitely agree that caching would be better but don't necessarily know how to go about implementing

Reply:
Good question - that's the difficult part :)

Firstly, I think it's OK for the cache to be in-process; it's casual usage that we're aiming to improve, and if a caller wants to perform specific robots-related crawl management, they're likely to be better off building something external to this library.

I wondered if we could store the parsed-robots-state as a _class attribute_ (not an instance attribute -- we want the cache to remain even after the caller uses a scraper for a few different pages).. but `wild_mode` [factory-builds](https://github.com/hhursev/recipe-scrapers/blob/356909007f9be4411e63773ac39cc49cf569febe/recipe_scrapers/__init__.py#L622) scraper classes on-demand, so it seems incompatible with that approach.

A simple albeit not particularly elegant solution would be to store an initially-empty dictionary alongside [`SCRAPERS`](https://github.com/hhursev/recipe-scrapers/blob/356909007f9be4411e63773ac39cc49cf569febe/recipe_scrapers/__init__.py#L290) that maps the `host_name` -- and this time we do want to use that instead of `url`, because it seems suitable as a derived key we can use for lookup -- and to refer to that when checking from `check_scraper_allowed`.

A caveat is that I think we'd want the cached values to expire after a reasonable amount of time.  That might seem inconsistent with the casual-interaction approach I'm considering here, but I do think it'd be valid for this code to be running within a desktop application where the process remains open and the user checks it once per day, for example.

Let me know what you think.

Reply:
Hmm... I do like the dictionary lookup idea, with `{host_name: last_scraper_check}` and then perhaps using the time module to see how long it's been since `last_scraper_check` if it's more than 24 hours (I think pinging robots.txt once a day would be a good starting point), then it would check robots.txt. Gonna need to let the idea 'cook' a bit more though...

Reply:
The python docs have a recipe that uses an OrderedDict to create a time bound least recently used cache: https://docs.python.org/3/library/collections.html#collections.OrderedDict:~:text=(key)-,An,-OrderedDict%20would%20also which looks like a good starting point.

Reply:
Hey, thank you for the contrib!
I want to keep this going WITHOUT doing the cache thing discussed.

I see no point in ad-hoc caching solution for the <domain>/robots.txt endpoint.

If we are to implement "request caching" mechanism I'd like it to work on all our requests. Including the ones fetching a recipe content. This should be done once we separate the resource fetching/networking from the scraping step in the project. So (long term) we'll end up with caching on/off, caching in memory/fs/redis, cache ttl, proxy on/off, etc. settings too on a more "global" level.

So, let's move this forward without implementing the cache thing.

I'll leave some other comments and change requests (mostly naming related) that I'd like to get addressed. They should be quick to do :fingers_crossed: